guest-verona-rt/arch/x86_64/hv_sev.cc

```diff
@@ -0,0 +1,598 @@
+  }
+
+  /**
+   * Heap setup is complex routine under SEV.
+   * Only source of memory information is the loader-injected unmeasured memory
+   * map. The memory map content should be viewed as byzantine. Also need to
+   * check the validity in relation to the known virtual top-of-memory. Abort if
+   * any inconsistency detected.
+   */
+  static void setup_heap_sev(void*)
+  {
+    // Compute the heap start, since the first memory range includes other
+    // elements as well.
+    uint8_t* heap_start = &__heap_start;
+
+    // This cannot be controlled by the hypervisor, but checking just in case
+    // the loader missed it.
+    if (
+      vmsa_settings->virtual_top_of_memory != 0 &&
+      (snmalloc::address_cast(heap_start) >=
+         vmsa_settings->virtual_top_of_memory ||
+       (vmsa_settings->virtual_top_of_memory % PAGE_SIZE) != 0))
+    {
+      kabort();
+    }
+
+    bool first_entry = true;
+    bool first_vtom_entry = true;
+    size_t last_entry_end_address = 0;
+    for (const auto& entry : unmeasured_loader_data->memory_map)
+    {
+      if (entry.is_null())
+      {
+        break;
+      }
+      size_t entry_address = entry.gpa_page_offset * HV_PAGE_SIZE;
+      size_t entry_size = entry.page_count * HV_PAGE_SIZE;
+      /**
+       * Validate entries using the following rules:
+       *  * The size cannot be 0.
+       *  * Ranges are non-overlapping and monotonically increasing.
+       *  * No range can overflow or end beyond 2^48.
+       *  * The first entry covers at least the heap start.
+       */
+      if (entry_size == 0)
+      {
+        kabort();
+      }
+      if (entry_address < last_entry_end_address)
+      {
+        kabort();
+      }
+      size_t current_entry_end_address = entry_address + entry_size;
+      if (
+        current_entry_end_address < entry_address ||
+        current_entry_end_address > (static_cast<uint64_t>(1) << 48))
+      {
+        kabort();
+      }
+      if (
+        first_entry &&
+        current_entry_end_address <= snmalloc::address_cast(heap_start))
+      {
+        kabort();
+      }
+      // Check for overlap with virtual top-of-memory and carve out visible
+      // range.
+      if (
+        vmsa_settings->virtual_top_of_memory != 0 &&
+        current_entry_end_address > vmsa_settings->virtual_top_of_memory)
+      {
+        if (first_vtom_entry)
+        {
+          size_t above_vtom_size =
+            current_entry_end_address - vmsa_settings->virtual_top_of_memory;
+          above_vtom_memory_range = std::span(
+            reinterpret_cast<uint8_t*>(vmsa_settings->virtual_top_of_memory),
+            above_vtom_size);
+          current_entry_end_address -= above_vtom_size;
+          entry_size -= above_vtom_size;
+          first_vtom_entry = false;
+        }
+        // Extra memory ranges beyond the top of the virtual top-of-memory are
+        // ignored.
+        else
+        {
+          entry_size = 0;
```

Comment:
I wonder if
```suggestion
          continue;
```
would be clearer.  And you move
```C++
last_entry_end_address = current_entry_end_address;
```
a bit earlier, e.g. line 416.

Reply:
I wasn't sure if I needed to use `last_entry_end_address` later on when building out the checks, but now that it is more final your proposal makes sense.